src/privileged/pageMonitor/api.js

```diff
@@ -0,0 +1,305 @@
+              Services.eTLD.getBaseDomainFromHost(e.data.telemetryData.origin);
+          } catch (error) {
+            return;
+          }
+          pageMonitorEventEmitter.emitPageBeforeUnload(tabId, e.data.telemetryData);
+        },
+        async pageDOMContentLoadedCallback(e) {
+          const tabId = tabTracker.getBrowserTabId(e.target);
+          try {
+            e.data.telemetryData.etld =
+              Services.eTLD.getBaseDomainFromHost(e.data.telemetryData.origin);
+          } catch (error) {
+            return;
+          }
+
+          pageMonitorEventEmitter.emitPageDOMContentLoaded(tabId, e.data.telemetryData);
+        },
+        async setListeners(win) {
+          const mm = win.getGroupMessageManager("browsers");
+          // We pass "true" as the third argument to signify that we want to listen
+          // to messages even when the framescript is unloading, to catch tabs closing.
+          mm.addMessageListener("CookieRestrictions:beforeunload", this.pageBeforeUnloadCallback, true);
+          mm.addMessageListener("CookieRestrictions:DOMContentLoaded", this.pageDOMContentLoadedCallback, true);
+          mm.addMessageListener("CookieRestrictions:pageError", this.pageErrorCallback);
+
+          mm.loadFrameScript(context.extension.getURL("privileged/pageMonitor/framescript.js"), true);
+
+          const shieldIcon = win.document.getElementById("tracking-protection-icon-box");
+          const trackingProtectionSection = win.document.getElementById("tracking-protection-container");
+          shieldIcon.style.display = "none";
+          trackingProtectionSection.style.display = "none";
+        },
+
+        async init(extensionSetExceptions) {
+          this.extensionSetExceptions = extensionSetExceptions;
+          EveryWindow.registerCallback("set-content-listeners", this.setListeners.bind(this), this.unmount.bind(this));
+
+          // Listen for addon disabling or uninstall.
+          AddonManager.addAddonListener(this);
+        },
+
+        async addException(currentOrigin) {
+          const recentWindow = getMostRecentBrowserWindow();
+          const tabId = tabTracker.getBrowserTabId(recentWindow.gBrowser.selectedBrowser);
+          const hasException = Services.perms.testExactPermissionFromPrincipal(recentWindow.gBrowser.contentPrincipal, "trackingprotection") === Services.perms.ALLOW_ACTION;
+          if (!hasException) {
+            const addExceptionButton = recentWindow.document.getElementById("tracking-action-unblock");
+            addExceptionButton.doCommand();
+            this.extensionSetExceptions.push(currentOrigin);
+            pageMonitorEventEmitter.emitExceptionSuccessfullyAdded(tabId);
+          }
+        },
+
+        removeExceptions(domains) {
+          for (const domain of domains) {
+            Services.perms.remove(Services.io.newURI(domain), "trackingprotection");
+          }
+        },
+
+        testPermission(url) {
```

Comment:
This doesn't use its `url` parameter, so if you're sure that the most recent browser window is what we always want to use here, you can probably fix up the consumer to avoid passing the url.

Reply:
👍